lib/posix-process/signal/deliver.c

```diff
@@ -0,0 +1,355 @@
+		UK_ASSERT(!(altstack->ss_flags & SS_ONSTACK));
+
+		altstack->ss_flags |= SS_ONSTACK;
+
+		uk_pr_debug("Using altstack, ss_sp: 0x%lx, ss_size: 0x%lx\n",
+			    (__u64)altstack->ss_sp, altstack->ss_size);
+
+		ulsp = ukarch_gen_sp(altstack->ss_sp, altstack->ss_size);
+	} else {
+		ulsp = usc->regs.rsp;
+		uk_pr_debug("Using the application stack @ 0x%lx\n", ulsp);
+	}
+
+	/* Signal handlers can make syscalls, that can invoke signal handlers,
+	 * that can make syscalls, and so on. We need to support nesting.
+	 * Since on syscall entry we switch to the auxp, nexted syscalls will
+	 * corrupt the previous syscall's stack. To avoid that from happening,
+	 * save the base auxsp of this syscall. Once the handler returns we will
+	 * restore the original value. Moreover, prepare a new base auxsp for
+	 * the nested syscall based on the current (aligned) value of the auxsp.
+	 */
+	saved_auxsp = uk_thread_current()->auxsp;
+
+	UK_ASSERT(IS_ALIGNED(saved_auxsp, UKARCH_ECTX_ALIGN));
+	UK_ASSERT(ukarch_read_sp() != saved_auxsp);
+
+	/* Assuming that we are operating on the auxsp and that
+	 * the sp has moved down since entry, align down to
+	 * UKARCH_ECTX_ALIGN so that a nested syscall finds the
+	 * auxsp aligned. Make sure we reserve enough space for
+	 * pprocess_signal_arch_jmp_handler() to save our general
+	 * purpose registers before we jump to the application
+	 * handler.
+	 */
+	uk_thread_current()->auxsp = ALIGN_DOWN(ukarch_read_sp() -
+						sizeof(struct __regs),
+						UKARCH_ECTX_ALIGN);
+	ukplat_lcpu_set_auxsp(uk_thread_current()->auxsp);
+
+	/* Switch to the user's sysregs before calling the signal handler */
+	sysregs = &usc->sysregs;
+	ukarch_sysregs_switch_ul(sysregs);
+
+	if (ks->ks_flags & SA_SIGINFO) {
+		pprocess_signal_arch_set_ucontext(usc, &ucontext);
+		pprocess_signal_arch_jmp_handler(sig->siginfo.si_signo,
+						 &sig->siginfo, &ucontext,
+						 ks->ks_handler, (void *)ulsp);
+		pprocess_signal_arch_get_ucontext(&ucontext, usc);
+	} else {
+		pprocess_signal_arch_jmp_handler(sig->siginfo.si_signo,
+						 NULL, NULL, ks->ks_handler,
+						 (void *)ulsp);
+	}
+
+	/* Switch back to the uk sysregs */
+	ukarch_sysregs_switch_uk(sysregs);
+
+	/* Restore the auxsp */
+	uk_thread_current()->auxsp = saved_auxsp;
+	ukplat_lcpu_set_auxsp(uk_thread_current()->auxsp);
+
+	if (ks->ks_flags & SA_ONSTACK) {
+		UK_ASSERT(altstack->ss_flags & SS_ONSTACK);
+		UK_ASSERT(!(altstack->ss_flags & SS_DISABLE));
+		altstack->ss_flags &= ~SS_ONSTACK;
+	}
+
+	uk_mutex_unlock(&this_process->signal->lock);
+}
+
+/* Deliver a signal to a thread. The caller must check that the signal
+ * is not masked by the handling thread or ignored by the process or its
+ * default disposition.
+ */
+static int do_deliver(struct posix_thread *pthread, struct uk_signal *sig,
+		      struct uk_syscall_ctx *usc)
+{
+	struct posix_process *pproc;
+	uk_sigset_t saved_mask;
+	int signum;
+
+	UK_ASSERT(sig && sig->siginfo.si_signo);
+	UK_ASSERT(pthread);
```

Comment:
`UK_ASSERT(pthread->signal);` and ditto for `pproc` below
Maybe also one for `usc`